feat: make ctrl-c during trade run show the route as-calculated

kfsone opened the pull request:
This feature allows you to interrupt a long running multi-hop route calculation without losing all that processing work...

```
Win|pwsh> python trade.py run --from shinrar/memorial --ly 150 --jumps 1 --hops 4 --cr 370m --cap 1088 --gpt 8000 --age 3 --pad l --fc=n --detail --detail --progress --prune-score=40 Searching for quality trades. This may take a few minutes. Please be patient. \ Scanning market data… rows 1,392,753  kept: buys 1,384,326, sells 418,838 | Scanning stations…  examined 1  kept 1
* Hop   1: .........1 origins
| origin 1/1  destinations checked: 1  best score: 0
* Hop   2: ........45 origins .. 9,095,680-27,890,880cr gain, 8,360-25,635cr/ton
/ origin 43/45  destinations checked: 3,537  best score: 27,688,173
NOTE: Pruned 299 origins
* Hop   3: .......450 origins .. 23,700,992-40,815,232cr gain, 10,892-18,757cr/ton
\ origin 31/450  destinations checked: 2,353  best score: 32,072,684
*** Ctrl-C Interrupt, stopping search...

############################################################################
User aborted route calculation.
############################################################################

Shinrarta Dezhra/Jameson Memorial -> Tionisla/Ing Ring (score: 68326125.735657)
  Load from Shinrarta Dezhra/Jameson Memorial (345ls, Pad:L, Plt:N, Flc:N, Ody:N, Shp:Y, Out:Y):
     1088 x Legal Drugs/Agronomic Treatment    2,860cr vs   15,427cr, <1 hr vs <1 hr
  Unload at Barnard's Star/Pohl Metallurgic Station (61ls, Pad:L, Plt:Y, Flc:N, Ody:Y, Shp:N, Out:N) => Gain 13,672,896cr (12,567cr/ton) => 383,672,896cr
  Load from Barnard's Star/Pohl Metallurgic Station (61ls, Pad:L, Plt:Y, Flc:N, Ody:Y, Shp:N, Out:N):
      639 x Consumer Items/Cryolite           10,510cr vs   29,428cr, <1 hr vs 1 hr
      449 x Consumer Items/Pyrophyllite        9,822cr vs   28,537cr, <1 hr vs 1 hr
  Unload at Puppis Sector UO-R b4-0/Democracy Space Station (4ls, Pad:L, Plt:N, Flc:N, Ody:N, Shp:Y, Out:Y) => Gain 20,491,637cr (18,834.2cr/ton) => 404,164,533cr
  Load from Puppis Sector UO-R b4-0/Democracy Space Station (4ls, Pad:L, Plt:N, Flc:N, Ody:N, Shp:Y, Out:Y):
     1088 x Legal Drugs/Agronomic Treatment      452cr vs   28,495cr, 1 hr vs 2 hrs
  Unload at Tionisla/Ing Ring (576ls, Pad:L, Plt:N, Flc:N, Ody:N, Shp:Y, Out:Y) => Gain 30,510,784cr (28,043cr/ton) => 434,675,317cr
  ----------------------------------------------------------------------------
Finish at Tionisla/Ing Ring (576ls, Pad:L, Plt:N, Flc:N, Ody:N, Shp:Y, Out:Y) gaining 64,675,317cr (19,814cr/ton) => est 434,675,317cr total
```

kfsone committed:
feat: make ctrl-c during trade run show the route as-calculated

This feature allows you to interrupt a long running multi-hop route calculation without
losing all that processing work...

```
Win|pwsh> python trade.py run --from shinrar/memorial --ly 150 --jumps 1 --hops 4 --cr 370m --cap 1088 --gpt 8000 --age 3 --pad l --fc=n --detail --detail --progress --prune-score=40
Searching for quality trades. This may take a few minutes. Please be patient.
\ Scanning market data… rows 1,392,753  kept: buys 1,384,326, sells 418,838
| Scanning stations…  examined 1  kept 1
* Hop   1: .........1 origins
| origin 1/1  destinations checked: 1  best score: 0
* Hop   2: ........45 origins .. 9,095,680-27,890,880cr gain, 8,360-25,635cr/ton
/ origin 43/45  destinations checked: 3,537  best score: 27,688,173
NOTE: Pruned 299 origins
* Hop   3: .......450 origins .. 23,700,992-40,815,232cr gain, 10,892-18,757cr/ton
\ origin 31/450  destinations checked: 2,353  best score: 32,072,684
*** Ctrl-C Interrupt, stopping search...

User aborted route calculation.

Shinrarta Dezhra/Jameson Memorial -> Tionisla/Ing Ring (score: 68326125.735657)
  Load from Shinrarta Dezhra/Jameson Memorial (345ls, Pad:L, Plt:N, Flc:N, Ody:N, Shp:Y, Out:Y):
     1088 x Legal Drugs/Agronomic Treatment    2,860cr vs   15,427cr, <1 hr vs <1 hr
  Unload at Barnard's Star/Pohl Metallurgic Station (61ls, Pad:L, Plt:Y, Flc:N, Ody:Y, Shp:N, Out:N) => Gain 13,672,896cr (12,567cr/ton) => 383,672,896cr
  Load from Barnard's Star/Pohl Metallurgic Station (61ls, Pad:L, Plt:Y, Flc:N, Ody:Y, Shp:N, Out:N):
      639 x Consumer Items/Cryolite           10,510cr vs   29,428cr, <1 hr vs 1 hr
      449 x Consumer Items/Pyrophyllite        9,822cr vs   28,537cr, <1 hr vs 1 hr
  Unload at Puppis Sector UO-R b4-0/Democracy Space Station (4ls, Pad:L, Plt:N, Flc:N, Ody:N, Shp:Y, Out:Y) => Gain 20,491,637cr (18,834.2cr/ton) => 404,164,533cr
  Load from Puppis Sector UO-R b4-0/Democracy Space Station (4ls, Pad:L, Plt:N, Flc:N, Ody:N, Shp:Y, Out:Y):
     1088 x Legal Drugs/Agronomic Treatment      452cr vs   28,495cr, 1 hr vs 2 hrs
  Unload at Tionisla/Ing Ring (576ls, Pad:L, Plt:N, Flc:N, Ody:N, Shp:Y, Out:Y) => Gain 30,510,784cr (28,043cr/ton) => 434,675,317cr
  ----------------------------------------------------------------------------
Finish at Tionisla/Ing Ring (576ls, Pad:L, Plt:N, Flc:N, Ody:N, Shp:Y, Out:Y) gaining 64,675,317cr (19,814cr/ton) => est 434,675,317cr total
```

review feedback

review feedback

- feat: add SimpleAbort exception for tidying up "expected error" output

SimpleAbort should only be intercepted at the top level of an application.

incorporated feedback review

- fixed typos,
- one way to describe ctrl-c, an exception type,
- leverage SimpleAbort and UserAbortedRun,
- allow Ctrl-C to be Ctrl-C if there are no results to preserve, it simplifies logic a lot

That's not the greatest user experience, but it's no worse than hitting ctrl-c already was, and has the advantage that if you do it when there are results, you get *something* rather than nothing.

Tromador commented:
Very cool idea. I like this a lot. One can tell that you are playing and using td as you are finding and addressing possible pain points and making qol stuff.

Couple of gotchas.

In getBestHops() you fall through to
```
if connections == 0:
    raise NoHopsError(...)
```
This creates an edge case where if the user presses ^c really quickly/early  (before connections += 1 has happened even once), you’ll raise NoHopsError and run_cmd will translate that into “No routes had reachable trading links…” rather than “aborted”.

I reckon you should only raise NoHopsError if not aborted, e.g. “connections==0 and not self.aborted”.

Similarly for early aborts, run_cmd does
```
routes = [Route(stations=(src,), hops=(), ... ) for src in cmdenv.origins]
```
Now if hop 0 is aborted before any hop results exist, newRoutes can be empty and your new early-break path triggers: 
```
if not newRoutes:
    if calc.aborted:
        results.summary.exception += "User aborted route calculation.\n"
        break
```
…but it doesn’t clear routes. So after the hop-loop, routes is still that seeded “start at origin” route list. render() will happily print it (it’ll render as Origin -> Origin with zero hops/gain), which is not what anyone wants.

Your later guard
```
if not routes:
    if calc.aborted:
        raise NoDataError("Route aborted before any results were found.")
```
then won't trigger because routes is not empty.

kfsone commented:
ooof - good catches. will address once I've caught up on wife points :)

eyeonus commented on tradedangerous/tradecalc.py:
```diff
@@ -1076,186 +1110,164 @@ def station_iterator(srcStation, origin_idx):
-                        if bool(tdenv.unique)
-                        else (d.system is goalSystem or d.distLy < srcGoalDist)
+            try:
+                if tdenv.debug >= 1:
```
https://github.com/eyeonus/Trade-Dangerous/blob/1b60caaee4657aed94fe70f84d227ffd3263b5d3/tradedangerous/tradeenv.py#L182-L196

Tromador commented:
You've obviously worked on the things I mentioned. I sincerely hope the below are from an excess of Xmas Cheer. (Not joking, hope you had a great time).

1: `no_routes_on_first_hop()` will crash for `--from` as a system (and for no `--from`)

At the top of the helper you have `start_system = cmdenv.origPlace.system.name()`. 
That explodes if:

- `cmdenv.origPlace` is `None` (no `--from` case), or
- `cmdenv.origPlace` is a `System` (because `Systems` don’t have `.system`).

Given `no_routes_on_first_hop()` is called whenever hop 0 produces no routes, this is a boom death moment.

2: Ctrl-C can still be misreported if `connections > 0` but `bestToDest` is still empty

In `getBestHops()` a Ctrl-C sets `self.aborted = True`. You only `return []` when `connections == 0`; otherwise you `break` and later build `newRoutes` from `bestToDest`.

But `connections` is incremented *before* we know we’ve got any viable trade (it happens before `items` / `fitFunction`). So you can hit Ctrl-C with `connections > 0` while `bestToDest` is still empty → `getBestHops()` yields `newRoutes == []` with `calc.aborted == True`.

Back in `run_cmd.run()`, the `if not newRoutes:` path for `hopNo > 0` still doesn’t check `calc.aborted`, so it drops into the normal failure reporting (`routeFailedRestrictions` / “SORRY: Could not find profitable destinations…”) and `break`s before the later `if calc.aborted:` guard is reached.

Net effect: user hits Ctrl-C, but we tell them “no profitable destinations” instead of “aborted”, and we may still print whatever routes existed from the previous hop (because `routes[:] = newRoutes` hasn’t happened yet).

3) Off-by-one in the “SORRY” message hop number

You changed it to:

- `... beyond hop #{hopNo:n}`

But hopNo is 0-based in that loop; the old message used `hopNo + 1`. This will print the wrong hop number in the user-facing output.

I also notice you've done some colour work in the same PR. I know you're already working frantically giving us tons of (despite my nit picking) cool code and that's really appreciated, but please can you keep the scope tight. Then we can merge what might be great, rather than holding it up for something unrelated which may have a problem.

Totally non-blocking nit picking stuff:

`Route.text(self, colorize: callable) — callable` is the builtin function, not a type. 
`sigmoid()` annotation still claims it might return `int` but it will always return `float` due to `/`.

kfsone commented:
I *appreciate* the nit-picking :) I'll scope out the non-ctrl-c stuff.

kfsone commented:
1- Ah - I hadn't moved the hoists into the correct block,
2- `if connections == 0` -> `if not bestToDest`, since that's what actually matters,
3- good catch

extracted the color code stuff to a separate diff

Tromador commented:
Please don't hate me:

1) `no_routes_on_first_hop()` still crashes for` --from` as a system

You’ve now gated it behind `if cmdenv.origPlace:`, which removes the `None` crash. Yay!

But you still `start_system = cmdenv.origPlace.system.name()`. If `cmdenv.origPlace` is a System this will still explode because System doesn’t have .system. 

So `--from 'Sol'` would still crash out.

2) Ctrl-C can still be misreported on `hop > 0`

Even with the `bestToDest` fix, the control-flow problem in run_cmd.run() is unchanged:

- Ctrl-C in `getBestHops()` can produce `newRoutes == []` and `calc.aborted == True`.
- For `hopNo > 0`, the `if not newRoutes:` block still does **not** check `calc.aborted`, so it can still fall into:
  - restriction failure messaging, or
  - `SORRY: Could not find profitable destinations...` and then breaks before reaching the later if `calc.aborted:` guard (which sits after `routes[:] = newRoutes`).

tl;dr:   user hits Ctrl-C mid-search and can still get told “no profitable destinations” instead of “aborted”.

kfsone commented:
Why would I hate you for solid code reviews, you lemon?

kfsone commented:
1- "System" has a ".system" property precisely so you can do `x.system` on either a system or a station
2- this code is gnarly, so I'm having a hard time wrapping my head around some of it :) I'm trying to ensure we provide lenience and don't tell you "didn't find anything" even tho it had found *some* stuff but hadn't gotten under-way evaluating a new set of next-hop routes

We're talking about this block, right?

<img width="803" height="432" alt="image" src="https://github.com/user-attachments/assets/8019e0f3-5ba4-4c17-8faf-46157bf3019c" />

Shouldn't pickedRoutes be true? (I'm v. unsure)

If so, then would it be

```
if pickedRoutes: break
if aborted: burp
```
?

Tromador commented:
> Why would I hate you for solid code reviews, you lemon?

LOL. In this case because I kept banging on about the same thing, but obviously I wasn't being serious.

.system stuff - yes, you're right. I seem to have a blindness about tradedb.py being a db, even if just in memory. I mean the clue is in the name. I think there might have been a problem originally, but it certainly looks ok in the current version.

Typo: Bort that run. Bort it good. `raise NoDataError(f"{ABORTED} borted run before any routes calculated.")`
Related pickiness: We now have three different ways of saying Aborted
- console print in `getBestHops()` (`"*** Ctrl-C Interrupt..."`)
- `ABORTED = "*** Ctrl+C: User aborted run"`
- and a summary exception appends `f"{ABORTED}.\n"` (note the extra period)

Actual bug: For `no_routes_on_first_hop()` messaging is wrong when `--start-jumps` was used

in `checkOrigins()` when `startJumps>0`  (--start-jumps)  there is
```
    cmdenv.origins = expandForJumps(tdb, cmdenv, cmdenv.origPlace, cmdenv.startJumps)
    cmdenv.origPlace = None
```
 
Which is fine, but when we get to `no_routes_on_first_hop()` 

```
if cmdenv.origPlace:
    ...
else:
    errText = "Could not find any trade links in the galaxy (no --from) with those criteria."
```

…but the user did supply --from; it was just cleared internally due to --start-jumps and it won't crash, but throws a bogus error.

Nitpick: `sigmoid()` annotated as `-> float | int` but always returns `float` because `/` produces float in py3.
Nitpick: `TradeDB,TradeEnv` in the TYPE_CHECKING import wants a space after the comma (stylistic maybe?)

kfsone commented:
Not passively-aggressive asking or anything, but just incase: I want to clarify I don't get offended if you make edits/changes. I mean, yes I do - in that "htf did I spell Aborted borted?" for a split second moment of embarrasment, in that competitive "I am perfect so if you can detect my flaws how much more perfect must you be, or maybe my original assumption I'm just a twonk was really correct" way, not in that "ok, I'm gonna need a whole lot more and a whole lot bigger hatchets" sort of way :)

Tromador commented:
> Not passively-aggressive asking or anything, but just incase: I want to clarify I don't get offended if you make edits/changes. 

No, I'm just dumb and don't know how to edit your PRs. It's that simple. I have to look up the most basic "how to do shit with github" and have made some truly epic screwups, trashing this repo which I then have to scrabble to fix before anyone realises that I've broken it. Again.

I have enough time dealing with my own anxiety disorder, I'm not trying to mess with anyone else's, or score stupid points over a typo. Life's too short. That's why I try (and maybe fail) to be light hearted or humorous in some of my comments. For me, just managing to get out of bed on any given day means I'm doing well with my party pack of physical and mental disabilities. There's no agenda beyond, "mate you made a typo". If that was the only problem, I'd probably have merged it and then fixed it and let it be rolled out down the line with some later fix.

kfsone commented:
You're in great standing with me, mate, I'm genuinely appreciating the reviews.

Tinkering on this right now, taking an exception based approach that I think might wrap this up.

kfsone commented:
LOL - I was trying to test the "couldn't find anything" path, only to discover it will happily plan routes in-system for you if you overly constrain it - I didn't realize it could do that.
<img width="950" height="347" alt="image" src="https://github.com/user-attachments/assets/427a7643-6ca4-417f-b131-1641bf6d2b34" />

kfsone commented:
Feedback addressed, and I couldn't resist tidying up the summary.exception presentation. Sorry :(

Interrupt after first hop
<img width="874" height="447" alt="image" src="https://github.com/user-attachments/assets/bd6f5276-a9ca-470b-abf7-b4e200da33ac" />

Interrupt before first hop has calculated any route (I added a "throw KeyboardInterrupt" before setting bestToDest[id] for this invocation


<img width="872" height="227" alt="image" src="https://github.com/user-attachments/assets/9c1ac115-910b-4296-9898-84e8ba6b3284" />

- Added a "SimpleAbort" exception we can use to say "you don't need a backtrace here", it's different than "TradeException" because it should only ever be intercepted at the top level.
- Added a "UserAbortedRun" exception to tradecalc so that we can be consistent across tools,
- Too lazy to leverage it beyond run for now.

Tromador commented:
I think this is really close:

`no_routes_on_first_hop()`still keys off `origPlace`, but `--start-jumps` clears `origPlace` after expanding origins, so the helper can emit the wrong generic message.

New corner-case: Ctrl-C very early in hop>1 can be treated as “before any routes calculated” and drop previously completed hop output because `getBestHops` re-raises when `bestToDest` is empty for that hop.

Tromador commented:
How about this? (I still don't know how to edit your PR).

```
def no_routes_on_first_hop(cmdenv: TradeEnv, calc: TradeCalc) -> None:
    """Handle the special case where run found no routes on the first hop."""
    # Is it because you ctrl-c'd?
    if calc.aborted:
        raise UserAbortedRun("during first hop before any routes found")

    # What did the user *start from* (as best as we can tell here)?
    start_place: str | None = cmdenv.starting or None
    start_system: str | None = None

    orig = getattr(cmdenv, "origPlace", None)
    if isinstance(orig, Station):
        # A concrete station start: we can give the most specific help.
        start_place = orig.name()
        start_system = orig.system.name()
    elif isinstance(orig, System):
        # System start: no station-specific suggestions.
        start_place = orig.name()
        start_system = orig.name()
    else:
        # --start-jumps (and friends) may expand origins and set origPlace=None,
        # but origSystems will still reflect what they started from.
        orig_systems = getattr(cmdenv, "origSystems", None)
        if orig_systems and len(orig_systems) == 1:
            start_system = orig_systems[0].name()
            if start_place is None:
                start_place = start_system

    max_ly = cmdenv.maxJumpsPer * cmdenv.maxLyPer
    exe = cmdenv.argv[0] if getattr(cmdenv, "argv", None) else sys.argv[0]

    if start_place:
        errText = f"No profitable buyers found for the goods at {start_place}.\n"
        if start_system:
            errText += (
                "\n"
                "You may want to try:\n"
                f"  {exe} local \"{start_system}\" --ly {max_ly} -vv --stations --trading"
            )
        # Only valid when we *know* the start was a station.
        if isinstance(orig, Station):
            errText += (
                "\n"
                "or:\n"
                f"  {exe} market \"{start_place}\" --sell -vv"
            )
    else:
        # No reliable starting context at all (i.e. genuinely not using --from/-f).
        errText = "Could not find any trade links in the galaxy with those criteria."

    raise NoDataError(errText)
```

kfsone committed:
review feedback

kfsone commented:
For the first part, there is a keyboardInterrupt handler around the *call* to getBestToDest, which sets newRoutes to [] and allows it to fall through.

<img width="632" height="535" alt="image" src="https://github.com/user-attachments/assets/e4c3b474-448a-439c-a22a-242f421dce05" />

and we can add a simple
```
@property
def system(self) -> System:
    return self
```
to the orm System, but we don't need to do battle with finding the system name here.

Submitting a new refactor of no routes on first hop, and I slightly changed the getBestHops wrap to care about routes, not pickedRoutes. 

Made it intentionally throw a KeyboardInterrupt *before* even calling getBestHops, and at various points during getBestHops:

```
(.venv) [12/30/2025 16:09:02 |dev/ctrl-c-run| S:\Elite\Trade-Dangerous]
Win|pwsh> python trade.py run --cr 700m --cap 1088 --ly 16 --link-ly 16 --jumps 2 --pad l --age 4 --ls-max 50000 --fc=n --hops=3 --gpt=12000 --supply 400 --demand 1000 --from "sol" --via sirius --progress -v -v --color --prune-score 40 --empty-ly 100 -w
🕸0: Command line was: ['trade.py', 'run', '--cr', '700m', '--cap', '1088', '--ly', '16', '--link-ly', '16', '--jumps', '2', '--pad', 'l', '--age', '4', '--ls-max', '50000', '--fc=n', '--hops=3', '--gpt=12000', '--supply', '400', '--demand', '1000', '--from', 'sol', '--via',
'sirius', '--progress', '-v', '-v', '--color', '--prune-score', '40', '--empty-ly', '100', '-w']
🕸0: reloadCache: ensure_fresh_db → kept
🕸0: persist data is stale by timestamp: cur=1767139764.0365222, file=1767139719.9353857
🕸0: Data load took 5.591s
Searching for quality trades. This may take a few minutes. Please be patient.
| Scanning market data… rows 1,667,377  kept: buys 1,222,698, sells 444,679
🕸0: origPlace: System: Sol
🕸0: filtering --from station list: Sol/Daedalus,Sol/Galileo,Sol/Abraham Lincoln,Sol/M.Gorbachev,Sol/Li Qing Jao,Sol/Mars High,Sol/Columbus,Sol/Titan City,Sol/Ehrlich City,Sol/Navarrete's Prospect,Sol/Majoro Entertainment Complex,Sol/Dewsnap Prospecting Station,Sol/Beauchene's
Horizon
🕸0: Using all available destinations
| Scanning stations…  examined 1  kept 1
🕸0: Via: {<System ID=121569805492 dbname='Sirius' pos=(6.25,-1.28125,-5.75)>}
🕸0: filtering --via station list: Sirius
* Hop   1: ........13 origins
| origin 1/13  destinations checked: 1  best score: 0
* Hop   2: ........81 origins .. 13,689,216-37,489,216cr gain, 12,582-34,457cr/ton
🕸0: ** Keyboard Interrupt
🕸0: ** User Aborted
#####################################################################
*** Ctrl+C: User aborted run: results may be incomplete or inaccurate
#####################################################################

Sol/Daedalus -> Sirius/Patterson Enterprise (score: 15442541.209242)
  Load from Sol/Daedalus (169ls, Pad:L, Plt:N, Flc:N, Ody:N, Shp:Y, Out:Y):
     1088 x Chemicals/Silver   34,426cr vs   48,441cr, <1 hr
  Unload at Sirius/Patterson Enterprise (1.04Kls, Pad:L, Plt:N, Flc:N, Ody:N, Shp:Y, Out:Y) => Gain 15,248,320cr (14,015cr/ton) => 715,248,320cr
  ----------------------------------------------------------------------------
Finish at Sirius/Patterson Enterprise (1.04Kls, Pad:L, Plt:N, Flc:N, Ody:N, Shp:Y, Out:Y) gaining 15,248,320cr (14,015cr/ton) => est 715,248,320cr total
```

If it throws after getBestHops (so newRoutes is populated but routes is just the first route, and we haven't yet done anything with the new routes):

```
Win|pwsh> python trade.py run --cr 700m --cap 1088 --ly 16 --link-ly 16 --jumps 2 --pad l --age 4 --ls-max 50000 --fc=n --hops=3 --gpt=12000 --supply 400 --demand 1000 --from "sol" --via sirius --progress -v -v --color --prune-score 40 --empty-ly 100
Searching for quality trades. This may take a few minutes. Please be patient.
| Scanning market data… rows 1,686,401  kept: buys 1,236,281, sells 450,120
| Scanning stations…  examined 1  kept 1
* Hop   1: ........13 origins
| origin 1/13  destinations checked: 1  best score: 0

*** Ctrl+C: User aborted run: before any routes calculated
```

if it did that in hopNo == 1:
```
Win|pwsh> python trade.py run --cr 700m --cap 1088 --ly 16 --link-ly 16 --jumps 2 --pad l --age 4 --ls-max 50000 --fc=n --hops=3 --gpt=12000 --supply 400 --demand 1000 --from "sol" --via sirius --progress -v -v --color --prune-score 40 --empty-ly 100
Searching for quality trades. This may take a few minutes. Please be patient.
\ Scanning market data… rows 1,658,305  kept: buys 1,216,848, sells 441,457
| Scanning stations…  examined 1  kept 1
* Hop   1: ........13 origins
| origin 1/13  destinations checked: 1  best score: 0
* Hop   2: ........81 origins .. 13,689,216-37,489,216cr gain, 12,582-34,457cr/ton
| origin 1/81  destinations checked: 1  best score: 0
#####################################################################
*** Ctrl+C: User aborted run: results may be incomplete or inaccurate
#####################################################################

Sol/Daedalus -> Sirius/Patterson Enterprise (score: 15442541.209242)
  Load from Sol/Daedalus (169ls, Pad:L, Plt:N, Flc:N, Ody:N, Shp:Y, Out:Y):
     1088 x Chemicals/Silver   34,426cr vs   48,441cr, <1 hr
  Unload at Sirius/Patterson Enterprise (1.04Kls, Pad:L, Plt:N, Flc:N, Ody:N, Shp:Y, Out:Y) => Gain 15,248,320cr (14,015cr/ton) => 715,248,320cr
  ----------------------------------------------------------------------------
Finish at Sirius/Patterson Enterprise (1.04Kls, Pad:L, Plt:N, Flc:N, Ody:N, Shp:Y, Out:Y) gaining 15,248,320cr (14,015cr/ton) => est 715,248,320cr total
```

Losing the last batch at the very last moment would suck a bit, but it's a race condition I don't think there's a real need to fight ahead of a larger overhaul of the way we're calculating the trades; it involves turning "getBestHops" that would involve turning getBestHops into a generator which would be a significant piece of work for an extreme edge case?

That scenario requires you hitting ctrl-c in the narrow window between getBestHops finishing and python leaving the except block; if you did it a microsecond later you'd just get a normal termination.

Addressing that would be an augmentation of this change rather than a necessary component.

So there's still a small window where hitting ctrl-c might lose you some calculations, but those are fairly egregious cases in the first place - like specifying --link-ly=250 --empty-ly=250 --start-jumps=100 --from colonia, and then hitting ctrl-c when it reaches 640,740/640,741 on hop 1 *and* getting super unlucky.

Whereas if you did this on hop 2, you'd still get the results from 0 and 1, which is still an improvement.

kfsone commented:
muahaha - appreciate your diligence, I can be a happy ctrl-cer now  :) I can be happy when I "craunch the marmoset"(*)


(* google for a laugh)
